File: patches/fix_app_quit.py
#!/usr/bin/env python3
"""
@patch-target: app.asar.contents/.vite/build/index.js
@patch-type: python

Fix app not quitting after cleanup completes.

After the will-quit handler calls preventDefault() and runs cleanup,
calling app.quit() again becomes a no-op on Linux. The will-quit event
never fires again, leaving the app stuck.

Solution: Use app.exit(0) instead of app.quit() after cleanup is complete.
Since all cleanup handlers have already run (mcp-shutdown, quick-entry-cleanup,
prototype-cleanup), we can safely force exit. Using setImmediate ensures
the exit happens in the next event loop tick.
"""

import sys
import re

def main():
    if len(sys.argv) != 2:
        print("Usage: fix_app_quit.py <file>")
        sys.exit(1)

    file_path = sys.argv[1]

    with open(file_path, 'rb') as f:
        content = f.read()

    print("=== Patch: fix_app_quit ===")
    print(f"  Target: {file_path}")

    # Original pattern: clearTimeout(n)}XX&&YY.app.quit()}
    # Variables change between versions (e.g., S_&&he → TS&&ce)
    # The XX&&YY.app.quit() doesn't work after preventDefault() on Linux
    # Replace with setImmediate + app.exit(0) for reliable exit
    pattern = rb'(clearTimeout\(\w+\)\})(\w+)&&(\w+)(\.app\.quit\(\))'

    def replacement(m):
        flag_var = m.group(2).decode('utf-8')
        electron_var = m.group(3).decode('utf-8')
        return f'{m.group(1).decode("utf-8")}if({flag_var}){{setImmediate(()=>{electron_var}.app.exit(0))}}'.encode('utf-8')

    new_content, count = re.subn(pattern, replacement, content)

    if count == 0:
        print("  [WARN] app.quit pattern: 0 matches (may need pattern update)")
        # Debug: check for any app.quit patterns
        if b'.app.quit()' in content:
            print("  [INFO] Found '.app.quit()' in file but pattern didn't match")
        sys.exit(0)

    print(f"  [OK] app.quit -> app.exit: {count} match(es)")

    with open(file_path, 'wb') as f:
        f.write(new_content)

    print("  [PASS] App quit patched successfully")
    sys.exit(0)

if __name__ == "__main__":
    main()

File: patches/fix_app_quit.nim
# @patch-target: app.asar.contents/.vite/build/index.js
# @patch-type: nim
#
# Fix app not quitting after cleanup completes.
#
# After the will-quit handler calls preventDefault() and runs cleanup,
# calling app.quit() again becomes a no-op on Linux. The will-quit event
# never fires again, leaving the app stuck.
#
# Solution: Use app.exit(0) instead of app.quit() after cleanup is complete.
# Since all cleanup handlers have already run (mcp-shutdown, quick-entry-cleanup,
# prototype-cleanup), we can safely force exit. Using setImmediate ensures
# the exit happens in the next event loop tick.

import std/[os, strutils]
import regex

proc apply*(input: string): string =
  # Original pattern: clearTimeout(n)}XX&&YY.app.quit()}
  # Variables change between versions (e.g., S_&&he -> TS&&ce)
  # Note: [\w$]+ is used because minified JS names can contain $ (e.g., f$, u$)
  # The XX&&YY.app.quit() doesn't work after preventDefault() on Linux
  # Replace with setImmediate + app.exit(0) for reliable exit
  let pattern = re2"(clearTimeout\([\w$]+\)\})([\w$]+)&&([\w$]+)(\.app\.quit\(\))"
  var count = 0
  result = input.replace(
    pattern,
    proc(m: RegexMatch2, s: string): string =
      inc count
      let grp0 = s[m.group(0)] # clearTimeout(n)}
      let flagVar = s[m.group(1)] # XX
      let electronVar = s[m.group(2)] # YY
      # group(3) is .app.quit() -- we discard it
      grp0 & "if(" & flagVar & "){setImmediate(()=>" & electronVar & ".app.exit(0))}",
  )
  if count == 0:
    if ".app.quit()" in input:
      echo "  [INFO] Found '.app.quit()' in file but pattern didn't match"
    echo "  [FAIL] app.quit pattern: 0 matches (may need pattern update)"
    quit(1)
  echo "  [OK] app.quit -> app.exit: " & $count & " match(es)"

when isMainModule:
  if paramCount() != 1:
    echo "Usage: fix_app_quit <file>"
    quit(1)
  let filePath = paramStr(1)
  echo "=== Patch: fix_app_quit ==="
  echo "  Target: " & filePath
  let input = readFile(filePath)
  let output = apply(input)
  writeFile(filePath, output)
  echo "  [PASS] App quit patched successfully"
